File: src/markdown.ts
// ── MarkdownV2 formatting ───────────────────────────────────────────────────

const MD2_SPECIAL = /([_*\[\]()~`>#+\-=|{}.!\\])/g

function escapeMd2(text: string): string {
  return text.replace(MD2_SPECIAL, "\\$1")
}

// Sentinel characters for placeholder substitution.
// Using Unicode private-use-area codepoints instead of raw control bytes
// to avoid collisions with any real content.
const PH = {
  IC_START: "\uE000", IC_END: "\uE001",
  LK_START: "\uE002", LK_END: "\uE003",
  BS: "\uE004", BE: "\uE005",
  IS: "\uE006", IE: "\uE007",
}

export function toMarkdownV2(text: string): string {
  const segments: string[] = []

  const codeBlockRe = /```(\w*)\n([\s\S]*?)```/g

  // Find all code blocks
  const blocks: { start: number; end: number; lang: string; code: string }[] = []
  let m: RegExpExecArray | null
  while ((m = codeBlockRe.exec(text)) !== null) {
    blocks.push({ start: m.index, end: m.index + m[0].length, lang: m[1]!, code: m[2]! })
  }

  // Build parts array
  const parts: { type: "text" | "codeblock"; content: string; lang?: string }[] = []
  let lastEnd = 0
  for (const block of blocks) {
    if (block.start > lastEnd) {
      parts.push({ type: "text", content: text.slice(lastEnd, block.start) })
    }
    parts.push({ type: "codeblock", content: block.code, lang: block.lang })
    lastEnd = block.end
  }
  if (lastEnd < text.length) {
    parts.push({ type: "text", content: text.slice(lastEnd) })
  }

  const inlineCodeRe = /`([^`\n]+)`/g

  for (const part of parts) {
    if (part.type === "codeblock") {
      const escaped = part.content.replace(/\\/g, "\\\\").replace(/`/g, "\\`")
      segments.push(`\`\`\`${part.lang ?? ""}\n${escaped}\`\`\``)
    } else {
      let t = part.content

      // Temporarily replace inline code
      const inlineCodes: string[] = []
      t = t.replace(inlineCodeRe, (_, code) => {
        const idx = inlineCodes.length
        const escaped = code.replace(/\\/g, "\\\\").replace(/`/g, "\\`")
        inlineCodes.push("`" + escaped + "`")
        return `${PH.IC_START}${idx}${PH.IC_END}`
      })

      // Temporarily replace links [text](url)
      const links: string[] = []
      t = t.replace(/\[([^\]]+)\]\(([^)]+)\)/g, (_, linkText, url) => {
        const idx = links.length
        links.push(`[${escapeMd2(linkText)}](${url.replace(/[)\\]/g, "\\$&")})`)
        return `${PH.LK_START}${idx}${PH.LK_END}`
      })

      // Convert **bold** → *bold* (Telegram uses single *)
      t = t.replace(/\*\*(.+?)\*\*/g, (_, inner) => `${PH.BS}${inner}${PH.BE}`)

      // Convert single *italic* → _italic_
      t = t.replace(/(?<!\*)\*(?!\*)(.+?)(?<!\*)\*(?!\*)/g, (_, inner) => `${PH.IS}${inner}${PH.IE}`)

      // Escape remaining special chars
      t = escapeMd2(t)

      // Restore placeholders
      t = t.replace(new RegExp(PH.BS, "g"), "*").replace(new RegExp(PH.BE, "g"), "*")
      t = t.replace(new RegExp(PH.IS, "g"), "_").replace(new RegExp(PH.IE, "g"), "_")

      const icRe = new RegExp(`${PH.IC_START}(\\d+)${PH.IC_END}`, "g")
      t = t.replace(icRe, (_, idx) => inlineCodes[Number(idx)]!)

      const lkRe = new RegExp(`${PH.LK_START}(\\d+)${PH.LK_END}`, "g")
      t = t.replace(lkRe, (_, idx) => links[Number(idx)]!)

      segments.push(t)
    }
  }

  return segments.join("")
}

export function thinkingToMd2(text: string): string {
  const escaped = escapeMd2(text)
  const lines = escaped.split("\n")
  if (lines.length === 0) return ""
  const result = lines.map((l, i) => {
    const prefix = i === 0 ? ">💭 " : ">"
    const suffix = i === lines.length - 1 ? "||" : ""
    return `${prefix}${l}${suffix}`
  })
  return result.join("\n")
}

export function splitMessage(text: string, limit: number): string[] {
  if (text.length <= limit) return [text]
  const chunks: string[] = []
  let remaining = text
  while (remaining.length > 0) {
    if (remaining.length <= limit) {
      chunks.push(remaining)
      break
    }
    let splitAt = remaining.lastIndexOf("\n", limit)
    if (splitAt < limit / 2) splitAt = limit
    chunks.push(remaining.slice(0, splitAt))
    remaining = remaining.slice(splitAt).trimStart()
  }
  return chunks
}

File: src/markdown.rs
use regex::Regex;

// MarkdownV2 special characters that need escaping
const MD2_SPECIAL: &str = r#"_*[]()~`>#+\-=|{}.!\\"#;

fn escape_md2(text: &str) -> String {
    let mut result = String::with_capacity(text.len() * 2);
    for ch in text.chars() {
        if MD2_SPECIAL.contains(ch) {
            result.push('\\');
        }
        result.push(ch);
    }
    result
}

// Unicode Private Use Area placeholders
const PH_IC_START: char = '\u{E000}';
const PH_IC_END: char = '\u{E001}';
const PH_LK_START: char = '\u{E002}';
const PH_LK_END: char = '\u{E003}';
const PH_BS: char = '\u{E004}';
const PH_BE: char = '\u{E005}';
const PH_IS: char = '\u{E006}';
const PH_IE: char = '\u{E007}';

/// Convert LLM markdown to Telegram MarkdownV2 format.
pub fn to_markdown_v2(text: &str) -> String {
    let code_block_re = Regex::new(r"```(\w*)\n([\s\S]*?)```").unwrap();

    // Find all code blocks
    let mut blocks: Vec<(usize, usize, String, String)> = Vec::new();
    for cap in code_block_re.captures_iter(text) {
        let m = cap.get(0).unwrap();
        blocks.push((
            m.start(),
            m.end(),
            cap.get(1).map_or("", |m| m.as_str()).to_string(),
            cap.get(2).map_or("", |m| m.as_str()).to_string(),
        ));
    }

    // Build parts
    let mut segments = Vec::new();
    let mut last_end = 0;

    for (start, end, lang, code) in &blocks {
        if *start > last_end {
            segments.push(convert_text_segment(&text[last_end..*start]));
        }
        // Code blocks: only escape ` and \
        let escaped = code.replace('\\', "\\\\").replace('`', "\\`");
        segments.push(format!("```{}\n{}```", lang, escaped));
        last_end = *end;
    }
    if last_end < text.len() {
        segments.push(convert_text_segment(&text[last_end..]));
    }

    segments.join("")
}

fn convert_text_segment(text: &str) -> String {
    let inline_code_re = Regex::new(r"`([^`\n]+)`").unwrap();
    let link_re = Regex::new(r"\[([^\]]+)\]\(([^)]+)\)").unwrap();
    let bold_re = Regex::new(r"\*\*(.+?)\*\*").unwrap();
    let italic_re = Regex::new(r"\*(.+?)\*").unwrap();

    let mut t = text.to_string();

    // Replace inline code with placeholders
    let mut inline_codes: Vec<String> = Vec::new();
    t = inline_code_re
        .replace_all(&t, |caps: &regex::Captures| {
            let idx = inline_codes.len();
            let code = caps.get(1).unwrap().as_str();
            let escaped = code.replace('\\', "\\\\").replace('`', "\\`");
            inline_codes.push(format!("`{}`", escaped));
            format!("{}{}{}", PH_IC_START, idx, PH_IC_END)
        })
        .to_string();

    // Replace links with placeholders
    let mut links: Vec<String> = Vec::new();
    t = link_re
        .replace_all(&t, |caps: &regex::Captures| {
            let idx = links.len();
            let link_text = caps.get(1).unwrap().as_str();
            let url = caps.get(2).unwrap().as_str();
            let escaped_url = url.replace(')', "\\)").replace('\\', "\\\\");
            links.push(format!("[{}]({})", escape_md2(link_text), escaped_url));
            format!("{}{}{}", PH_LK_START, idx, PH_LK_END)
        })
        .to_string();

    // Convert **bold** to Telegram *bold*
    t = bold_re
        .replace_all(&t, |caps: &regex::Captures| {
            format!("{}{}{}", PH_BS, caps.get(1).unwrap().as_str(), PH_BE)
        })
        .to_string();

    // Convert *italic* to Telegram _italic_
    t = italic_re
        .replace_all(&t, |caps: &regex::Captures| {
            format!("{}{}{}", PH_IS, caps.get(1).unwrap().as_str(), PH_IE)
        })
        .to_string();

    // Escape remaining special chars
    t = escape_md2(&t);

    // Restore placeholders
    t = t.replace([PH_BS, PH_BE], "*");
    t = t.replace([PH_IS, PH_IE], "_");

    // Restore inline codes
    let ic_re = Regex::new(&format!(
        "{}(\\d+){}",
        regex::escape(&PH_IC_START.to_string()),
        regex::escape(&PH_IC_END.to_string())
    ))
    .unwrap();
    t = ic_re
        .replace_all(&t, |caps: &regex::Captures| {
            let idx: usize = caps.get(1).unwrap().as_str().parse().unwrap();
            inline_codes.get(idx).cloned().unwrap_or_default()
        })
        .to_string();

    // Restore links
    let lk_re = Regex::new(&format!(
        "{}(\\d+){}",
        regex::escape(&PH_LK_START.to_string()),
        regex::escape(&PH_LK_END.to_string())
    ))
    .unwrap();
    t = lk_re
        .replace_all(&t, |caps: &regex::Captures| {
            let idx: usize = caps.get(1).unwrap().as_str().parse().unwrap();
            links.get(idx).cloned().unwrap_or_default()
        })
        .to_string();

    t
}

/// Format thinking/reasoning text as MarkdownV2 expandable blockquote.
pub fn thinking_to_md2(text: &str) -> String {
    let escaped = escape_md2(text);
    let lines: Vec<&str> = escaped.lines().collect();
    if lines.is_empty() {
        return String::new();
    }
    lines
        .iter()
        .enumerate()
        .map(|(i, line)| {
            let prefix = if i == 0 { ">💭 " } else { ">" };
            let suffix = if i == lines.len() - 1 { "||" } else { "" };
            format!("{}{}{}", prefix, line, suffix)
        })
        .collect::<Vec<_>>()
        .join("\n")
}

/// Split a message into chunks that fit within Telegram's character limit.
pub fn split_message(text: &str, limit: usize) -> Vec<String> {
    if text.len() <= limit {
        return vec![text.to_string()];
    }
    let mut chunks = Vec::new();
    let mut remaining = text;
    while !remaining.is_empty() {
        if remaining.len() <= limit {
            chunks.push(remaining.to_string());
            break;
        }
        let split_at = remaining[..limit]
            .rfind('\n')
            .filter(|&pos| pos >= limit / 2)
            .unwrap_or(limit);
        chunks.push(remaining[..split_at].to_string());
        remaining = remaining[split_at..].trim_start();
    }
    chunks
}

/// Escape text for safe embedding in XML-like channel wrapper.
pub fn sanitize_for_xml(text: &str) -> String {
    text.replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
}
